// cells-semaas-adapter/src/connector.js
function request(method) {
  return async (options) => {
    const { url, body, headers } = options;
    const configuration = {
      method,
      body: JSON.stringify(body),
      headers: {
        'content-type': 'application/json',
      },
    };

    if (headers) {
      Object.keys(headers).forEach((header) => {
        const headerValue = headers[header];

        configuration.headers[header] = headerValue;
      });
    }

    // OJO!!!
    // Interceptar aquí errores
    const response = await fetch(url, configuration);
    let result = {};
    try {
      result = await response.json();
    } catch (e) {
      // empty server response...couldn't be proccessed but still OK
    }

    if (!response.ok && result && result.status !== 400) {
      return null; 
    }

    return result;
  };
}

const METHODS = {
  GET: request('get'),
  POST: request('post'),
};

export default class Connector {
  static async get(url, headers) {
    return METHODS.GET({ url, headers });
  }

  static async post(url, body, headers) {
    return METHODS.POST({ url, body, headers });
  }
}
